api/graphql/mutations/cohort.py

```diff
@@ -0,0 +1,171 @@
+    async def create_cohort_from_criteria(
+        self,
+        info: Info,
+        cohort_spec: CohortBodyInput,
+        cohort_criteria: CohortCriteriaInput | None = None,
+        dry_run: bool = False,
+    ) -> Annotated['GraphQLCohort', strawberry.lazy('api.graphql.schema')]:
+        """
+        Create a cohort with the given name and sample/sequencing group IDs.
+        """
+        from api.graphql.schema import GraphQLCohort
+
+        connection: Connection = info.context['connection']
+        connection.check_access(
+            {ProjectMemberRole.writer, ProjectMemberRole.contributor}
+        )
+        clayer = CohortLayer(connection)
+
+        if not connection.project:
+            raise ValueError('A cohort must belong to a project')
+
+        if not cohort_criteria and not cohort_spec.template_id:
+            raise ValueError(
+                'A cohort must have either criteria or be derived from a template'
+            )
+
+        internal_project_ids: list[ProjectId] = []
+
+        if cohort_criteria and cohort_criteria.projects:
+            projects = connection.get_and_check_access_to_projects_for_names(
+                project_names=cohort_criteria.projects, allowed_roles=ReadAccessRoles
+            )
+
+            internal_project_ids = [p.id for p in projects if p.id]
+
+        template_id_raw = (
+            cohort_template_id_transform_to_raw(cohort_spec.template_id)
+            if cohort_spec.template_id
+            else None
+        )
+        assert connection.project_id
+        cohort_output = await clayer.create_cohort_from_criteria(
+            project_to_write=connection.project_id,
+            description=cohort_spec.description,
+            cohort_name=cohort_spec.name,
+            dry_run=dry_run,
+            cohort_criteria=(
+                CohortCriteria.from_dict(
+                    strawberry.asdict(cohort_criteria)
+                ).to_internal(internal_project_ids)
+                if cohort_criteria
+                else None
+            ),
+            template_id=template_id_raw,
+        )
+        created_cohort = (
+            await clayer.query(
+                CohortFilter(id=GenericFilter(eq=cohort_output.cohort_id))
+            )
+        )[0]
+
+        return GraphQLCohort.from_internal(created_cohort)
+
+    @strawberry.mutation
+    async def create_cohort_template(
+        self,
+        template: CohortTemplateInput,
+        info: Info,
+    ) -> Annotated['GraphQLCohortTemplate', strawberry.lazy('api.graphql.schema')]:
+        """
+        Create a cohort template with the given name and sample/sequencing group IDs.
+        """
+        from api.graphql.schema import GraphQLCohortTemplate
+
+        connection: Connection = info.context['connection']
+        connection.check_access(
+            {ProjectMemberRole.writer, ProjectMemberRole.contributor}
+        )
+        clayer = CohortLayer(connection)
+
+        if not connection.project:
```

Comment:
This will always error out as the connection won't have a project, the same applies to all uses of `connection.project` and `connection.project_id` in this file.